### PR TITLE
gh-60283: Check for redefined test names in CI

### LINKED ISSUE #505: Detect duplicate method names in CI
> [...] Maybe we can just use pyflakes in a CI job, only looking for this specific kind of issue?

_Originally posted by @vstinner in https://github.com/python/cpython/issues/105560#issuecomment-1609308613_

### LINKED ISSUE #60283: list duplicate test names with patchcheck
BPO | [16079](https://bugs.python.org/issue16079)
--- | :---
Nosy | @rhettinger, @gpshead, @vstinner, @rbtcollins, @ned-deily, @ezio-melotti, @merwok, @cjerdonek, @xdegaye, @ambv, @miss-islington
PRs | <li>python/cpython#12827</li><li>python/cpython#12828</li><li>python/cpython#12886</li><li>python/cpython#12940</li><li>python/cpython#12950</li>
Files | <li>[duplicate_test_names.patch](https://bugs.python.org/file27326/duplicate_test_names.patch "Uploaded as text/plain at 2012-09-28.10:15:36 by @xdegaye")</li><li>[duplicate_code_names.py](https://bugs.python.org/file27375/duplicate_code_names.py "Uploaded as text/plain at 2012-10-01.16:49:48 by @xdegaye")</li><li>[std_lib_duplicates.txt](https://bugs.python.org/file27376/std_lib_duplicates.txt "Uploaded as text/plain at 2012-10-01.16:50:30 by @xdegaye")</li><li>[duplicate_code_names_2.py](https://bugs.python.org/file31891/duplicate_code_names_2.py "Uploaded as text/plain at 2013-09-28.15:28:24 by @xdegaye")</li><li>[ignored_duplicates](https://bugs.python.org/file31892/ignored_duplicates "Uploaded as text/plain at 2013-09-28.15:28:47 by @xdegaye")</li><li>[duplicate_code_names_3.py](https://bugs.python.org/file48265/duplicate_code_names_3.py "Uploaded as text/plain at 2019-04-13.16:34:39 by @xdegaye")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2012-09-28.10:15:37.662>
labels = ['easy', 'tests', 'type-bug', 'library', '3.9']
title = 'list duplicate test names with patchcheck'
updated_at = <Date 2019-11-27.18:45:45.910>
user = 'https://github.com/xdegaye'
```

bugs.python.org fields:
```python
activity = <Date 2019-11-27.18:45:45.910>
actor = 'brett.cannon'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)', 'Tests']
creation = <Date 2012-09-28.10:15:37.662>
creator = 'xdegaye'
dependencies = []
files = ['27326', '27375', '27376', '31891', '31892', '48265']
hgrepos = []
issue_num = 16079
keywords = ['patch', 'easy']
message_count = 31.0
messages = ['171428', '171499', '171505', '171509', '171511', '171512', '171513', '171519', '171536', '171544', '171548', '171558', '171559', '171576', '171734', '198525', '198586', '229625', '340168', '340219', '340221', '340252', '340286', '340295', '340323', '340584', '340585', '340687', '340785', '340786', '351668']
nosy_count = 11.0
nosy_names = ['rhettinger', 'gregory.p.smith', 'vstinner', 'rbcollins', 'ned.deily', 'ezio.melotti', 'eric.araujo', 'chris.jerdonek', 'xdegaye', 'lukasz.langa', 'miss-islington']
pr_nums = ['12827', '12828', '12886', '12940', '12950']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue16079'
versions = ['Python 3.9']
```

</p></details>

### Linked PRs
* gh-109161

### PR DESCRIPTION
**Pull request by hugovk**
Fixes https://github.com/python/core-workflow/issues/505.
Fixes https://github.com/python/cpython/issues/60283.

Add Ruff to the CI (via pre-commit) to check `Lib/test/` for [F811 "Redefinition of unused name"](https://beta.ruff.rs/docs/rules/redefined-while-unused/). Ruff (47s) is faster than Flake8 (1m39s).

This will help find copy/pasted `test_*` method names, which results in one of them not being run.

It also finds redefined variables, which ISTR people were in favour of. 

There are some files which fail to lint or parse, and have been excluded. Some of these are intentional and should never be checked (e.g. `Lib/test/badsyntax_pep3120.py`), some are due to new 3.12+ syntax not yet supported by the Ruff linter (e.g. `Lib/test/test_type_aliases.py`). I think we can be pragmatic and exclude any that add new syntax until such a time the linter supports the new version. (This may also help drive adoption of new syntax by the linter.) 

This currently fails due to duplicated tests in `test_monitoring`, which will be fixed in https://github.com/python/cpython/pull/109139 and has been left as a demonstration in the meantime.

All other `test_*` duplicates have been fixed (see links in https://github.com/python/core-workflow/issues/505). 

Some files have duplicate variables, I've excluded those and they can be fixed and re-included in the future.

* Issue: gh-60283

### CODE REVIEW

File: .ruff.toml
Comment (AlexWaygood) [inline .ruff.toml:1]: How extensible is this approach? I've been meaning for a while to open an issue/PR about running pyflakes (or the equivalent ruff lints) on `Tools/clinic/` in CI. But we'd (I'd) probably want more lints enabled on `Tools/clinic/` than we're either able to or would want to enable on the whole of `Lib/test/`

File: .pre-commit-config.yaml
Comment (AlexWaygood) [inline .pre-commit-config.yaml]: What does the autofix do for F811? I'm slightly suspicious of this, as you don't always fix this issue the same way. Sometimes you really do have two identical methods, and you just need to delete one. But sometimes one of the identical methods just needs to be renamed.

File: .ruff.toml
Comment (hugovk) [reply .ruff.toml:1]: I'd been wondering the same thing, so tested it out. 

And the good news is we can have a config file per directory!

For example:

# `.pre-commit-config.yaml`

```yaml
repos:
  - repo: https://github.com/astral-sh/ruff-pre-commit
    rev: v0.0.287
    hooks:
      - id: ruff
        args: [--fix, --exit-non-zero-on-fix]
        files: ^Lib/test/

  - repo: https://github.com/astral-sh/ruff-pre-commit
    rev: v0.0.287
    hooks:
      - id: ruff
        args: [--fix, --exit-non-zero-on-fix]
        files: ^Tools/
```

# `Lib/test/.ruff.toml`

```toml
select = [
    "F811",
]
extend-exclude = [
    # Failed to lint
    "badsyntax_pep3120.py",
    "encoded_modules/module_iso_8859_1.py",
    "encoded_modules/module_koi8_r.py",
    # Failed to parse
    "mod_generics_cache.py",
    "badsyntax_3131.py",
    "support/socket_helper.py",
    "test_builtin.py",
    "test_fstring.py",
    "test_funcattrs.py",
    "test_functools.py",
    "test_type_aliases.py",
    "test_type_params.py",
    "test_typing.py",
    # TODO Fix: F811 Redefinition of unused name
    "test__opcode.py",
    "test_buffer.py",
    "test_ctypes/test_arrays.py",
    "test_ctypes/test_functions.py",
    "test_dataclasses/__init__.py",
    "test_descr.py",
    "test_enum.py",
    "test_genericclass.py",
    "test_grammar.py",
    "test_import/__init__.py",
    "test_keywordonlyarg.py",
    "test_pkg.py",
    "test_subclassinit.py",
    "test_unittest/testmock/testpatch.py",
    "test_yield_from.py",
    "time_hashlib.py",
]
```

# `Tools/.ruff.toml`

```toml
select = [
    "E401",
]
extend-exclude = [
    # Failed to lint
    "i18n/pygettext.py",
]
```

# Run it

```console
❯ pre-commit run --all-files ruff
ruff.....................................................................Failed
- hook id: ruff
- exit code: 1

Lib/test/test_binascii.py:279:9: F811 Redefinition of unused `test_hex_roundtrip` from line 236
Lib/test/test_monitoring.py:1211:9: F811 Redefinition of unused `test_line_then_instruction` from line 1188
Lib/test/test_monitoring.py:1216:9: F811 Redefinition of unused `test_instruction_then_line` from line 1193
Found 3 errors.

ruff.....................................................................Failed
- hook id: ruff
- exit code: 1

Tools/c-analyzer/distutils/ccompiler.py:6:1: E401 Multiple imports on one line
Tools/c-analyzer/distutils/msvccompiler.py:11:1: E401 Multiple imports on one line
Tools/c-analyzer/distutils/unixccompiler.py:16:1: E401 Multiple imports on one line
Tools/c-analyzer/distutils/util.py:86:9: E401 Multiple imports on one line
Tools/freeze/checkextensions_win32.py:25:1: E401 Multiple imports on one line
Tools/freeze/freeze.py:432:9: E401 Multiple imports on one line
Tools/freeze/winmakemakefile.py:1:1: E401 Multiple imports on one line
Tools/unicode/gencjkcodecs.py:1:1: E401 Multiple imports on one line
Tools/unicode/gencodec.py:29:1: E401 Multiple imports on one line
Tools/unicode/listcodecs.py:9:1: E401 Multiple imports on one line
Found 10 errors.
```

So I'll already move this PR's config file to `Lib/test`. 👍

File: .pre-commit-config.yaml
Comment (hugovk) [reply .pre-commit-config.yaml]: Nothing for F811, because as you mention, it's not safe to delete one.

The rules with autofix are labelled with 🛠 at https://beta.ruff.rs/docs/rules/

I added it more for the future, but shall we remove it until then?

File: .ruff.toml
Comment (AlexWaygood) [reply .ruff.toml:1]: Very nice, that looks like a perfect solution!

File: .pre-commit-config.yaml
Comment (AlexWaygood) [reply .pre-commit-config.yaml]: I'm in favour of autofixes in general, but I think it's nice for people to be able to see easily from our config file which rules we expect to be autofixed where possible, and which rules we don't want autofixes for (even if they exist). We could maybe do something like this?

```diff
diff --git a/.pre-commit-config.yaml b/.pre-commit-config.yaml
index 1d83c9414d..3cc332ab34 100644
--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -3,7 +3,7 @@ repos:
     rev: v0.0.287
     hooks:
       - id: ruff
-        args: [--fix, --exit-non-zero-on-fix]
+        args: [--exit-non-zero-on-fix]
         files: ^Lib/test/

   - repo: https://github.com/pre-commit/pre-commit-hooks
diff --git a/Lib/test/.ruff.toml b/Lib/test/.ruff.toml
index 1a381d84de..55619536d4 100644
--- a/Lib/test/.ruff.toml
+++ b/Lib/test/.ruff.toml
@@ -1,6 +1,10 @@
+fix = true
 select = [
     "F811",
 ]
+unfixable = [
+    "F811",
+]
 extend-exclude = [
     # Failed to lint
     "badsyntax_pep3120.py",
```

File: Lib/test/.ruff.toml
Comment (AlexWaygood) [inline Lib/test/.ruff.toml]:
```suggestion
    "F811",  # Redefinition of unused variable (useful for finding test methods with the same name)
```

File: .pre-commit-config.yaml
Comment (hugovk) [reply .pre-commit-config.yaml]: Hmm, I'm not sure about explicitly listing out all the fixer-less rules as unfixable, we'd be duplicating a lot. And some rules might grow a fixer later (e.g. it's not been implemented yet), and I think that would be the time to evaluate whether to disable it.

We're already being cautious and starting with just a single rule. And when we do have fixers, it is always up to the human to review and commit the changes. (And hopefully with a second human reviewer for the PR!)

I do think moving `fix = true` to the Ruff config makes sense, it'll be picked up when running Ruff as a CLI (outside pre-commit) too. I'll add this.

What do you think?

Comment (AlexWaygood) [reply .pre-commit-config.yaml]: Sure, I see the point about duplication. Looks pretty good to me now, thanks!

**Review by AlexWaygood — APPROVED** [review]
LGTM, thanks! I guess we just have to wait for #109139 to be merged first?

File: .pre-commit-config.yaml
Comment (AA-Turner) [inline .pre-commit-config.yaml:7]: How about adding ``--diff`` and ``--format github``?

The former will write the diff of the proposed fix to the logs, and the latter should output check annotations for GitHub for PRs

```suggestion
        args: ['--exit-non-zero-on-fix', '--diff', '--format', 'github']
```

Comment (AA-Turner) [reply .pre-commit-config.yaml:7]: Ref: https://beta.ruff.rs/docs/configuration/#command-line-interface

Comment (hugovk) [reply .pre-commit-config.yaml:7]: Hmm, I added them (re-ordered, without quotes, and combined `--format, github` into a single argument) but Ruff didn't fail as expected?

Expected:
* https://github.com/python/cpython/actions/runs/6130878966/job/16640392499?pr=109161#step:4:126

Actual:
* https://github.com/python/cpython/actions/runs/6136692710/job/16651782392?pr=109161#step:4:126
* https://github.com/hugovk/cpython/actions/runs/6136706649/job/16651813966?pr=109161#step:4:126

Comment (hugovk) [reply .pre-commit-config.yaml:7]: Caused by `--diff`, I've removed it.

Comment (hugovk) [reply .pre-commit-config.yaml:7]: With `--format=github`:

```console
❯ pre-commit run --all-files ruff
ruff.....................................................................Failed
- hook id: ruff
- exit code: 1

::error title=Ruff (F811),file=/Users/hugo/github/cpython/Lib/test/test_binascii.py,line=279,col=9,endLine=279,endColumn=27::Lib/test/test_binascii.py:279:9: F811 Redefinition of unused `test_hex_roundtrip` from line 236
::error title=Ruff (F811),file=/Users/hugo/github/cpython/Lib/test/test_monitoring.py,line=1211,col=9,endLine=1211,endColumn=35::Lib/test/test_monitoring.py:1211:9: F811 Redefinition of unused `test_line_then_instruction` from line 1188
::error title=Ruff (F811),file=/Users/hugo/github/cpython/Lib/test/test_monitoring.py,line=1216,col=9,endLine=1216,endColumn=35::Lib/test/test_monitoring.py:1216:9: F811 Redefinition of unused `test_instruction_then_line` from line 1193
```

The annotations are good for PRs:

<details>
<summary>GitHub screenshot</summary>

image

</details>

Without `--format=github` (not shown here, filename is bold, F811 is red):

```console
❯ pre-commit run --all-files ruff
ruff.....................................................................Failed
- hook id: ruff
- exit code: 1

Lib/test/test_binascii.py:279:9: F811 Redefinition of unused `test_hex_roundtrip` from line 236
Lib/test/test_monitoring.py:1211:9: F811 Redefinition of unused `test_line_then_instruction` from line 1188
Lib/test/test_monitoring.py:1216:9: F811 Redefinition of unused `test_instruction_then_line` from line 1193
Found 3 errors.
```

<details>
<summary>Console screenshot</summary>

image


</details>


It would be nice to have the GitHub formatting just for PRs and not for local use. 

Is it possible to apply `--format=github` just via the workflow? Like with an env var or something?

**Comment by hugovk** [comment]
Should we have changes to `.pre-commit-config.yaml` and `.ruff.toml` NOT trigger the main test build?

That would mean adding to this grep pattern:

https://github.com/python/cpython/blob/2dd6a86c4ee604b331ed739c2508b0d0114993c6/.github/workflows/build.yml#L66

**Comment by AlexWaygood** [comment]
> Should we have changes to `.pre-commit-config.yaml` and `.ruff.toml` NOT trigger the main test build?

That sounds like a good idea to me!

File: .pre-commit-config.yaml
Comment (hugovk) [reply .pre-commit-config.yaml:7]: Yes, `RUFF_FORMAT=github` should do it:
```
--format <FORMAT>
    Output serialization format for violations [env: RUFF_FORMAT=] [possible values: text, json, json-lines, junit, grouped, github, gitlab, pylint, azure]
```

Comment (AlexWaygood) [reply .pre-commit-config.yaml:7]: The annotations on this PR look great -- this was a fab idea @AA-Turner!

File: .pre-commit-config.yaml
Comment (AlexWaygood) [inline .pre-commit-config.yaml:7]: Make it clearer that the job is scoped to just `Lib/test/` (this will also help differentiate it from other ruff jobs targeting different parts of the repo, if we add them in the future)

```suggestion
        name: Run Ruff on Lib/test/
        args: [--exit-non-zero-on-fix]
```

https://pre-commit.com/#config-name

Comment (hugovk) [reply .pre-commit-config.yaml:7]: (👍 and edited to capitalise "Ruff")

**Review by AlexWaygood — APPROVED** [review]

**Comment by hugovk** [comment]
> LGTM, thanks! I guess we just have to wait for #109139 to be merged first?

Yes, or exclude `test_monitoring` here and re-include it #109139. Although it's been zero weekdays since #109139 was opened, so we can wait a bit longer.

File: .github/workflows/lint.yml
Comment (AA-Turner) [inline .github/workflows/lint.yml:10]: Whilst we're here?
```suggestion
  FORCE_COLOR: 1
  RUFF_FORMAT: github
```

**Review by AA-Turner — APPROVED** [review]
Thanks!

A

File: Lib/test/.ruff.toml
Comment (AlexWaygood) [inline Lib/test/.ruff.toml:3]: > > LGTM, thanks! I guess we just have to wait for #109139 to be merged first?
> 
> Yes, or exclude `test_monitoring` here and re-include it #109139.

This got me thinking, "Hmm... or we could just `# noqa` the problematic tests for now", which then got me thinking, "Hmm... if there's a possibility that we might have `# noqa` comments in `Lib/test/`, maybe we should add some validations for those in CI?"

```suggestion
    "F811",  # Redefinition of unused variable (useful for finding test methods with the same name)
    "RUF100",  # Unused noqa comment
    "PGH004",  # don't use blanket noqa comments; only ignore specific rules
```

Comment (AA-Turner) [reply Lib/test/.ruff.toml:3]: PGH004 I'd be less sure about, as it ties NoQA comments to Ruff's format. For example I believe support is being introduced for the more readable pylint-like style codes, but these might then sneak in over the "standard" pycodestyle codes.

Comment (AlexWaygood) [reply Lib/test/.ruff.toml:3]: I've no strong opinion, but I don't think there's really a defined standard for `noqa` comments. It used to be "whatever pycodestyle/flake8 support", since it _used_ to be that these comments were specific to those tools. Now it's more like "whatever pycodestyle/flake8 support if you're using those tools, OR whatever pylint supports if you're using that tool, OR whatever ruff supports if you're using _that_ tool". We're using ruff in CI, not any of those other tools, and it makes sense to me to tie the noqa comments we'd be using to the tool we're using them for?

Comment (AA-Turner) [reply Lib/test/.ruff.toml:3]: Fair enough, I suppose the concern was potential churn/lock-in when the next great tool comes around, but not a blocker in any real sense.

File: .github/workflows/lint.yml
Comment (hugovk) [reply .github/workflows/lint.yml:10]: pre-commit already enables colour for its logs, but we might as well have some nice colour for pip as well :)

File: Lib/test/.ruff.toml
Comment (hugovk) [reply Lib/test/.ruff.toml:3]: I generally think these are a good idea, we have the Flake8 equivalents for Pillow.

But let's leave it out of this PR because there's quite a few warnings to fix:


<details>
<summary>PGH004 and RUF100 warnings</summary>

```
Lib/test/test_asyncio/test_events.py:2885:22: PGH004 Use specific rule codes when using `noqa`
Lib/test/test_asyncio/test_events.py:2885:22: RUF100 [*] Unused blanket `noqa` directive
Lib/test/test_asyncio/test_runners.py:182:60: PGH004 Use specific rule codes when using `noqa`
Lib/test/test_asyncio/test_runners.py:182:60: RUF100 [*] Unused blanket `noqa` directive
Lib/test/test_binascii.py:279:9: F811 Redefinition of unused `test_hex_roundtrip` from line 236
Lib/test/test_context.py:454:19: PGH004 Use specific rule codes when using `noqa`
Lib/test/test_context.py:454:19: RUF100 [*] Unused blanket `noqa` directive
Lib/test/test_inspect.py:2798:37: PGH004 Use specific rule codes when using `noqa`
Lib/test/test_inspect.py:2798:37: RUF100 [*] Unused blanket `noqa` directive
Lib/test/test_inspect.py:2801:54: PGH004 Use specific rule codes when using `noqa`
Lib/test/test_inspect.py:2801:54: RUF100 [*] Unused blanket `noqa` directive
Lib/test/test_inspect.py:2806:43: PGH004 Use specific rule codes when using `noqa`
Lib/test/test_inspect.py:2806:43: RUF100 [*] Unused blanket `noqa` directive
Lib/test/test_inspect.py:2810:44: PGH004 Use specific rule codes when using `noqa`
Lib/test/test_inspect.py:2810:44: RUF100 [*] Unused blanket `noqa` directive
Lib/test/test_monitoring.py:1211:9: F811 Redefinition of unused `test_line_then_instruction` from line 1188
Lib/test/test_monitoring.py:1216:9: F811 Redefinition of unused `test_instruction_then_line` from line 1193
Lib/test/test_tomllib/burntsushi.py:11:116: RUF100 [*] Unused `noqa` directive (non-enabled: `E501`)
Lib/test/test_tomllib/burntsushi.py:12:104: RUF100 [*] Unused `noqa` directive (non-enabled: `E501`)
Lib/test/test_tomllib/burntsushi.py:22:20: RUF100 [*] Unused `noqa` directive (non-enabled: `C901`)
Lib/test/test_zipfile/__main__.py:3:27: RUF100 [*] Unused `noqa` directive (non-enabled: `F401`)
Found 21 errors.
[*] 11 potentially fixable with the --fix option.
```

</details>

And we'd have to decide _how_ to fix these, because we can see, for example, Ruff wants to strip this F401 "unused import" warning:

```diff
-from . import load_tests  # noqa: F401
+from . import load_tests
```

But we do "need" it, because it would normally be warned, even though Ruff has been configured to ignore F401 so thinks it's unused.

Comment (AlexWaygood) [reply Lib/test/.ruff.toml:3]: Good points. And various core devs often run pyflakes manually on the `Lib/` dir on a periodic basis in order to find unused imports. So these codes shouldn't be enabled for now, agreed!

**Comment by AlexWaygood** [comment]
🥳

**Comment by hugovk** [comment]
For the backport PRs, I updated the exclude lists, as they were slightly different.

And interestingly, we don't need to explicitly exclude the "Failed to lint/parse" files because they're printed to screen but don't fail the command.

For example, these all have return code = 0 = success:

```console
❯ pre-commit run --all-files ruff
Run Ruff on Lib/test/....................................................Passed
```
```console
❯ pre-commit run --all-files ruff --verbose
Run Ruff on Lib/test/....................................................Passed
- hook id: ruff
- duration: 0.02s

error: Failed to parse Lib/test/tokenizedata/badsyntax_3131.py:2:1: Got unexpected token €
error: Failed to parse Lib/test/test_fstring.py:671:28: f-string: expecting '}'
warning: Failed to lint Lib/test/badsyntax_pep3120.py: stream did not contain valid UTF-8
warning: Failed to lint Lib/test/encoded_modules/module_koi8_r.py: stream did not contain valid UTF-8
warning: Failed to lint Lib/test/encoded_modules/module_iso_8859_1.py: stream did not contain valid UTF-8
error: Failed to parse Lib/test/support/socket_helper.py:300:33: f-string: expecting '}'
```
```console
❯ ruff Lib/test
warning: Failed to lint Lib/test/encoded_modules/module_iso_8859_1.py: stream did not contain valid UTF-8
warning: Failed to lint Lib/test/encoded_modules/module_koi8_r.py: stream did not contain valid UTF-8
warning: Failed to lint Lib/test/badsyntax_pep3120.py: stream did not contain valid UTF-8
error: Failed to parse Lib/test/support/socket_helper.py:300:33: f-string: expecting '}'
error: Failed to parse Lib/test/test_fstring.py:671:28: f-string: expecting '}'
error: Failed to parse Lib/test/tokenizedata/badsyntax_3131.py:2:1: Got unexpected token €
```

Shall we explicitly exclude them (which makes the verbose output cleaner, and in which case I'll update the backports) or keep the default behaviour (in which case I'll update `main`)?

**Comment by AlexWaygood** [comment]
I think it's better to explicitly exclude  the unlintable files, as we do on `main`. It's maybe slightly harder to maintain, but the improvement in UI is worth it if contributors don't see all those warnings printed to the terminal 👍

**Comment by erlend-aasland** [comment]
Thanks for doing this!

**Pull request merged**
